ocaml/xapi/xapi_session.ml

```diff
@@ -787,6 +787,67 @@ module Caching = struct
 
   module AuthenticationCache : EXTERNAL_AUTH_CACHE =
     Helpers.AuthenticationCache.Make (String) (AuthenticatedResult)
+
+  let cache = ref None
+
+  let lock = Mutex.create ()
+
+  let ( let@ ) = ( @@ )
+
+  (* Attain the extant cache or get nothing if caching is
+     disabled. This function exists to delay the construction of the
+     cache, as Xapi_globs configuration is not guaranteed to have been
+     populated before the top-level code of this module is executed. *)
+  let get_or_init_cache () =
+    if not !Xapi_globs.external_authentication_cache_enabled then
+      None
+    else
+      let capacity = !Xapi_globs.external_authentication_cache_size in
+      Mutex.lock lock ;
```

Comment:
We have a Threadext.Mutex.execute that might be useful: it ensures that the lock is always released, even if something raises an exception.